DAOS-13102 control: Update MD-on-SSD server config validation rules

tanabarr opened the pull request:
Update validation rules for server config file regarding bdev_roles in
bdev storage tiers. Remove implicit role assignments for all storage
tier configurations so that MD-on-SSD will only ever be enabled if
bdev_roles are set explicitly in server config file engine storage
tiers.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

daosbuild1 commented on src/control/server/storage/config.go:
```diff
@@ -299,134 +299,123 @@
-//   - A role can only be assigned to one entire tier, i.e. tier 2 & 3 cannot both be assigned the
-//     Meta role. This doesn’t apply to the Data role which can be applied to multiple tiers e.g.
-//     in the case where > 3 bdev tiers exist.
-//   - All roles (WAL, Meta and Data) need to be assigned in bdev tiers if scm class is ram.
-//   - If the (first) scm tier is of class dcpm, then only one bdev tier with Data role is supported,
-//     no third tier (for now).
-func (tcs TierConfigs) validateBdevTierRoles() error {
-	sc := tcs.ScmConfigs()[0]
+// - No storage tier with class: dcpm exists, exactly one storage tier with class:
+//   ram exists, and zero or one tier(s) with class: nvme exists. If an NVMe
+//   tier is present, no bdev_roles: attributes are specified for the NVMe tier.
+//   This is the traditional DRAM-based (ephemeral) configuration, and it shall not
+//   use MD-on-SSD (to be compatible/consistent with earlier levels of DAOS software).
+//
+// - No storage tier with class: dcpm exists, exactly one storage tier with class:
+//   ram exists, and one, two or three tiers with class: nvme exist, with mandatory
+//   bdev_roles: attributes on each of the NVMe tiers. Each of the three roles
+//   (wal,meta,data) must be assigned to exactly one NVMe tier (no default
+//   assignments of roles to tiers by the control plane software; all roles shall be
+//   explicitly specified). This setup shall use the MD-on-SSD code path.  In this
+//   scenario allow the use of a single NVMe tier co-locating all three roles, three
+//   separate NVMe tiers with each tier dedicated to exactly one role, or two
+//   separate NVMe tiers where two of the three roles are co-located on one of the
+//   two NVMe tiers. In the latter case, all combinations to co-locate two of the
+//   roles shall be allowed, although not all those combinations may be technically
+//   desirable in production environments.
```
```suggestion
//   - Exactly one storage tier with class: dcpm exists, no storage tier with class:
//     ram exists, and zero or one tier(s) with class: nvme exists. If an NVMe tier is
//     present, no bdev_roles: attributes are allowed on it.  This is the traditional
//     PMem-based configuration, obviously not using MD-on-SSD.
//
//   - No storage tier with class: dcpm exists, exactly one storage tier with class:
//     ram exists, and zero or one tier(s) with class: nvme exists. If an NVMe
//     tier is present, no bdev_roles: attributes are specified for the NVMe tier.
//     This is the traditional DRAM-based (ephemeral) configuration, and it shall not
//     use MD-on-SSD (to be compatible/consistent with earlier levels of DAOS software).
//
//   - No storage tier with class: dcpm exists, exactly one storage tier with class:
//     ram exists, and one, two or three tiers with class: nvme exist, with mandatory
//     bdev_roles: attributes on each of the NVMe tiers. Each of the three roles
//     (wal,meta,data) must be assigned to exactly one NVMe tier (no default
//     assignments of roles to tiers by the control plane software; all roles shall be
//     explicitly specified). This setup shall use the MD-on-SSD code path.  In this
//     scenario allow the use of a single NVMe tier co-locating all three roles, three
//     separate NVMe tiers with each tier dedicated to exactly one role, or two
//     separate NVMe tiers where two of the three roles are co-located on one of the
//     two NVMe tiers. In the latter case, all combinations to co-locate two of the
//     roles shall be allowed, although not all those combinations may be technically
//     desirable in production environments.
```

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/1/execution/node/145/log

tanabarr commented:
Converted to draft because there is still pending discussion on design changes between @mjmac and @Michael-Hennecke .

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Unit Test on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11978/3/testReport/(root)/

kjacque commented on src/control/server/storage/bdev/backend_json.go:
```diff
@@ -221,6 +221,7 @@ func getSpdkConfigMethods(req *storage.BdevWriteConfigRequest) (sscs []*SpdkSubs
 			f = getAioKdevCreateMethod
 		}
 
+		fmt.Printf("roles tier %+v", tier)
```
Probably need to remove this before landing, right?

kjacque commented on src/control/server/storage/faults.go:
```diff
@@ -60,15 +60,23 @@ func FaultScmInvalidPMem(msg string) *fault.Fault {
+var FaultScmConfigTierMissing = storageFault(
+	code.ScmConfigTierMissing,
+	"missing scm storage tier in engine storage config",
+	"add a scm tier at index-0 in engine storage tiers list in server config file and restart daos_server")
```
Just a thought for clarity, instead of saying "at index 0," how about "in the first position"? The indices may not be obvious to a user writing a config file.

tanabarr replied:
done

kjacque commented on src/control/server/storage/config.go:
```diff
@@ -299,64 +299,120 @@ func (tcs TierConfigs) HaveBdevRoleMeta() bool {
-		return FaultBdevConfigBadNrRoles("WAL", foundWALTiers, nrWALTiers)
+
+	// MD-on-SSD configurations supports 1, 2 or 3 bdev tiers.
+	if len(bcs) > 3 {
```
Would be best to make 3 a constant

tanabarr replied:
done

kjacque reviewed (commented):
My comments are all pretty minor. Overall looks good.

knard-intel commented on src/control/server/storage/faults.go:
```diff
@@ -60,15 +60,23 @@ func FaultScmInvalidPMem(msg string) *fault.Fault {
-	"Install PMem modules and retry command")
-
-// FaultBdevConfigTypeMismatch represents an error where an incompatible mix of emulated and
+var FaultScmNoModules = storageFault(
```
Nit, the following error name could be more comprehensible:
```suggestion
var FaultScmNoPmemModules = storageFault(
```

tanabarr replied:
done

knard-intel commented on src/control/server/storage/config.go:
```diff
@@ -299,64 +299,120 @@ func (tcs TierConfigs) HaveBdevRoleMeta() bool {
+		roles := bc.Bdev.DeviceRoles
+		if roles.IsEmpty() {
+			if hasRoles() {
+				return FaultBdevConfigRolesMissing
```
From what I  understand, if the first bdev has no roles and the second one has some then the `FaultBdevConfigRolesMissing` will not be detected ?

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/5/execution/node/145/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage NLT on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11978/8/testReport/(root)/

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/9/execution/node/1014/log

knard-intel reviewed (dismissed):
LGTM

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/10/execution/node/1032/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/11/execution/node/1014/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11978/12/testReport/(root)/

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/13/execution/node/246/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/13/execution/node/211/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/13/execution/node/222/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/13/execution/node/219/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/15/execution/node/1033/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/16/execution/node/217/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/16/execution/node/220/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/16/execution/node/214/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/17/execution/node/184/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/17/execution/node/215/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/17/execution/node/228/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/18/execution/node/225/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/18/execution/node/228/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/18/execution/node/222/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/19/execution/node/194/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11978/20/testReport/(root)/

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/21/execution/node/647/log

tanabarr commented:
In CI run for this PR: https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-11978/21/tests/ there were 2*daos_racer failures. They are unrelated to this PR and same failure is seen on weekly master: https://build.hpdd.intel.com/job/daos-stack/job/daos/job/weekly-testing/lastCompletedBuild/testReport/FTEST_daos_racer/DaosRacerTest/Test___Functional_Hardware_Large___1___daos_racer_parallel_py_DaosRacerTest_test_parallel_run_daos_racer_hosts_job_manager_server .

This PR will need to be re-based after PR-11944 and PR-11797 lands as it is based on the latter and will cause failures without the former as it removes implicit validation. When re-basing it should just require the normal PR test flag as a sanity check.

After PR lands a weekly test run should be triggered to verify the operation after all 3 PRs have landed and role assignment is explicit and scm_size calculation is automatic.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/22/execution/node/1209/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/22/execution/node/1160/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/22/execution/node/1254/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

knard-intel commented:
If yes, it seems that it will not be any more possible to store the control plane metadata into the SCM RAM-FS defined by the scm_mount field.
Indeed, when I explicitly define in the `control_metadata` with the same value I got the following error.
```
$ dmg storage format
ERROR: Errors:
  Hosts    Error
  -----    -----
  wolf-196 pbin: code = 5 description = "write spdk nvme config: create: open /mnt/daos0/daos_control/engine0/daos_nvme.conf: no such file or directory"
```

Formatting is  properly done with the same daos server configuration when I just update the  `control_metadata:path` field from `/mnt/daos0` to `/tmp/daos_server/`

tanabarr commented:
> It seems that it will not be any more possible to store the control plane metadata into the SCM RAM-FS defined by the `scm_mount` field. Indeed, when I explicitly define in the `control_metadata` with the same value I got the following error.
> 
> ```
> $ dmg storage format
> ERROR: Errors:
>   Hosts    Error
>   -----    -----
>   wolf-196 pbin: code = 5 description = "write spdk nvme config: create: open /mnt/daos0/daos_control/engine0/daos_nvme.conf: no such file or directory"
> ```
> 
> Formatting is properly done with the same daos server configuration when I just update the `control_metadata:path` field from `/mnt/daos0` to `/tmp/daos_server/`

This is unexpected, I will look into it thank you @knard-intel .

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/25/execution/node/948/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/25/execution/node/1042/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/26/execution/node/1228/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/26/execution/node/1249/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/26/execution/node/1384/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/27/execution/node/1233/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/27/execution/node/1342/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/27/execution/node/1249/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/31/execution/node/1297/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/31/execution/node/1260/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/31/execution/node/1276/log

knard-intel reviewed (approved):
LGTM

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/32/execution/node/672/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/32/execution/node/812/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/32/execution/node/719/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/33/execution/node/703/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/33/execution/node/812/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

tanabarr commented:
@kjacque @knard-intel @mjmac reviewers: I had to force push to fix the list of commits after Phils control_metadata path CI fix landed to feature/vos_on_blob branch, sorry for inconvenience. The only change since last review is the last commit which merges in another PR to fix some bugs and needs the rest of this PR as a prerequisite.

GATEKEEPER, please use the PR title and description as commit message when merging into the feature branch.

tanabarr commented:
build pushed with parameters https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11978/39/parameters/

tanabarr commented:
flake8 and pylint check failures seem unrelated to this PR

phender commented:
> flake8 and pylint check failures seem unrelated to this PR

These are due to changes I made in a test util file in the feature/vos_on_blob branch and a similar function Dalton added on the master branch.  We're resolving the problem.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/39/execution/node/1270/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/39/execution/node/1286/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11978/39/execution/node/1421/log

knard-intel commented on src/control/server/ctl_storage_rpc.go:
```diff
@@ -421,25 +476,70 @@ func (c *ControlService) StorageFormat(ctx context.Context, req *ctlpb.StorageFo
-		if msg, hasError := instanceErrored[ei.Index()]; hasError {
+	for idx, ei := range instances {
+		if msg, hasError := instanceErrors[idx]; hasError {
 			c.log.Errorf("instance %d: %s", ei.Index(), msg)
```
Nit at line 543 s/ec.Index/idx/

tanabarr replied:
done

knard-intel commented on src/control/server/server_utils.go:
```diff
@@ -502,13 +502,32 @@ func cleanEngineHugepages(srv *server) error {
-// ensuring reported available memory (of type RAM) is enough to cover at least 90% of the
-// combined total memory of all engine RAM-disk sizes set in the storage config.
+// ensuring reported available memory (of type RAM) is enough to cover at least 90% of engine
+// AM-disk sizes set in the storage config.
```
s/AM-disk/RAM-disk/

knard-intel reviewed (approved):
Just some minor comments.

tanabarr commented:
> @tanabarr
> 
> * FTEST_daos_test.DAOS_Extend_Simple
>   
>   * I think https://daosio.atlassian.net/browse/DAOS-13412
> * daos_test/rebuild.py
>   
>   * This ticket is a different issue, but same test - https://daosio.atlassian.net/browse/DAOS-13279
> * pool/create_capacity.py
>   
>   * I don't see a ticket, but this test is notorious

thanks a lot @daltonbohning , I have updated the comment above

kjacque reviewed (approved):
Just a minor comment, otherwise lgtm.

tanabarr commented on src/control/server/ctl_storage_rpc.go:
```diff
@@ -325,92 +327,145 @@ func (c *ControlService) StorageScan(ctx context.Context, req *ctlpb.StorageScan
+		scmCfgs[idx] = scmCfg
+	}
+
+	if allNeedFormat {
+		// Check available RAM is sufficient before formatting SCM on engines.
+		if err := checkTmpfsMem(req.log, scmCfgs, req.getMemInfo); err != nil {
+			return nil, nil, err
```
A separate check is performed on engine start https://github.com/daos-stack/daos/pull/11978/files#diff-d4337e9b4e14476c4e9211f64d51adcdfc20fe8f35ce486e450063d637ef1d7cR530 that validates there is enough memory to support the single engine's ram-disk. This check on first format validates that there will be enough for ram-disks of all engines whereas when we do the per-engine check we don't have visibility on the ram-disk(s) of any other engines.

tanabarr replied:
but yes, this could be modified to verify any of the subsets being formatted, I will keep that in mind for future refactors. In MD-on-SSD phase 2 we will probably get rid of the concept of SCM format when in MD-on-SSD mode and just perform the check on start. The only problem then is that when checking available memory we need to be looking at the capacity usage of each engine's tmpfs to make sure we have enough available memory to cover all the unused capacities.

kjacque replied:
I see. Thanks for the thorough explanation!
